MatrixNMS-8 spec. (#5948)
openvinotoolkit/openvino: docs/ops/sort/MatrixNMS_8.md

> ##  MatrixNms<a name="MatrixNms"></a> {#openvino_docs_ops_sort_MatrixNms_8}

Reviewer: Update `ie_docs.xml` file as in the https://github.com/openvinotoolkit/openvino/pull/5773

> ##  MatrixNms<a name="MatrixNms"></a> {#openvino_docs_ops_sort_MatrixNms_8}

Reviewer: Let's rename operation to `MatrixNonMaxSuppression` to be consistent with already existing `NonMaxSuppression` operation.

**Versioned name**: *MatrixNms-8*

**Category**: *Sorting and maximization*

**Short description**: *MatrixNms* performs matrix non maximum suppression (NMS) of the boxes with predicted scores.

**Detailed description**: *MatrixNms* is a multi-phase operation.

> First it selects candidate bounding boxes that have higher scores than `score_threshold`. Then for each class at most `nms_top_k` number of candidate boxes are selected. Score of the candidate boxes are then decayed according to Matrix NMS algorithm [Wang et al](https://arxiv.org/abs/2003.10152.pdf).This algorithm is applied independently to each class and each batch element. Boxes of `background_class` are skipped and thus eliminated during the process. Finanly, it thresholds the decayed scores with `post_threshold`, and selects at most `keep_top_k` scoring candidate boxes per batch element.

Reviewer: Please, split this text into several lines of about 120 characters. That will be easier to modify/review in the future.

> First it selects candidate bounding boxes that have higher scores than `score_threshold`. Then for each class at most `nms_top_k` number of candidate boxes are selected. Score of the candidate boxes are then decayed according to Matrix NMS algorithm [Wang et al](https://arxiv.org/abs/2003.10152.pdf).This algorithm is applied independently to each class and each batch element. Boxes of `background_class` are skipped and thus eliminated during the process. Finanly, it thresholds the decayed scores with `post_threshold`, and selects at most `keep_top_k` scoring candidate boxes per batch element.

Reviewer: Missing space character in this part of the text `2003.10152.pdf).This`.

Author: Not sure exactly where is the missing space. Is it between "pdf)" and "This"? That's the only difference I find compared to NonMaxSuppression spec.

Reviewer: Yes, add a space before "This"

> First it selects candidate bounding boxes that have higher scores than `score_threshold`. Then for each class at most `nms_top_k` number of candidate boxes are selected. Score of the candidate boxes are then decayed according to Matrix NMS algorithm [Wang et al](https://arxiv.org/abs/2003.10152.pdf).This algorithm is applied independently to each class and each batch element. Boxes of `background_class` are skipped and thus eliminated during the process. Finanly, it thresholds the decayed scores with `post_threshold`, and selects at most `keep_top_k` scoring candidate boxes per batch element.

Reviewer: "Finanly, it thresholds the decayed scores with `post_threshold`" -> "Finally, the algorithm selects boxes with the decayed scores higher than `post_threshold`"

> The Matrix NMS algorithm is described as below:

Reviewer: remove "as"

> 1.  Compute `n x n` pairwise IoU matrix `X` for top N bounding boxes. Suppose `n` is the number of candidate boxes, which have been sorted descending by score, thresholded with `score_threshold` and then seleted with `nms_top_k`.

Reviewer: `seleted` -> `selected`

> 1.  Compute `n x n` pairwise IoU matrix `X` for top N bounding boxes. Suppose `n` is the number of candidate boxes, which have been sorted descending by score, thresholded with `score_threshold` and then seleted with `nms_top_k`.

Reviewer: Expand "IoU" for the first time. Then use the abbreviation.

> 1.  Compute `n x n` pairwise IoU matrix `X` for top N bounding boxes. Suppose `n` is the number of candidate boxes, which have been sorted descending by score, thresholded with `score_threshold` and then seleted with `nms_top_k`.

Author: resolved

> 2.  `triu`: Set the lower triangle and diagonal of `X` to 0.

Reviewer: It is not clear what "triu" means

> 3.  Take th column-wise max of `X` to compute a vector `K` of maximun IoU for each candidate box.

Reviewer: "th" -> "the"

4.  Repeat element value of `K` along axis 1. Suppose this get the matrix `X_cmax`.
5.  Compute the decay factor: `decay_factor = exp((X_cmax**2 - X**2) * gaussian_sigma)` if `decay_function` is `guassian`, else `decay_factor = (1 - X) / (1 - X_cmax)`.
6.  Take the column-wise min of `decay_factor`, and element-wise multiply with scores to decay them.

**Attributes**:

* *sort_result*

>   * **Description**: *sort_result* specifiy order of output elements.

Reviewer: "specifiy" -> "specify the"

  * **Range of values**: `class`, `score`, `none`

>     * *class* - sort selected boxes by class id (ascending) in each batch element.

Reviewer: Let's add additional attribute "*sort_result_descending*" as in the NonMaxSuppression operation which defines the sorting order.

    * *score* - sort selected boxes by score (descending) in each batch element.
    * *none* - do not guarantee the order in each batch element.
  * **Type**: `string`
  * **Default value**: `none`
  * **Required**: *No*

* *output_type*

>   * **Description**: the output tensor type

Reviewer: It is not clear which output this attribute affects. For example, the first output with `[class_id, box_score, xmin, ymin, xmax, ymax]` cannot be of type `i64` and `i32` for sure. Please, clarify this and consider renaming the attribute to be more precise which output it affects.

  * **Range of values**: `i64` or `i32`
  * **Type**: `string`
  * **Default value**: `i32`
  * **Required**: *No*

* *score_threshold*

  * **Description**: minimum score to consider box for the processing.
  * **Range of values**: a floating-point number
  * **Type**: `float`
  * **Default value**: `0`
  * **Required**: *No*

* *nms_top_k*

  * **Description**: maximum number of boxes to be selected per class.
  * **Range of values**: an integer starting from `-1`.
  * **Type**: `int`
  * **Default value**: `-1` meaning to keep all boxes
  * **Required**: *No*

* *keep_top_k*

  * **Description**: maximum number of boxes to be selected per batch element.
  * **Range of values**: an integer starting from `-1`.
  * **Type**: `int`
  * **Default value**: `-1` meaning to keep all boxes
  * **Required**: *No*

* *background_class*

  * **Description**: the background class id.

>   * **Range of values**: an integer from `-1` to the maximum class id

Reviewer: Are classes counted from 0? If so, then the maximum value is `max_class_id - 1`.

Author: It is tricky here. class id starts from 0. And PaddlePaddle uses -1 meaning keep all classes (no background class). The maximum class id should not be -1. It comes from shape of input scores. So maybe we can specify any integer out of [0, num_class] to mean keep all classes.

  * **Type**: `int`
  * **Default value**: `-1` meaning to keep all classes
  * **Required**: *No*

* *decay_function*

  * **Description**: decay function used to decay scores.
  * **Range of values**: `gaussian`, `linear`
  * **Type**: `string`
  * **Default value**: `linear`
  * **Required**: *No*

* *gaussian_sigma*

  * **Description**: gaussian_sigma parameter for gaussian decay_function.
  * **Range of values**: a floating-point number
  * **Type**: `float`
  * **Default value**: `2.0`
  * **Required**: *No*

* *post_threshold*

>   * **Description**: threshold to filter out boxes with low confidence score `after` decaying.

Reviewer: Why "after" is wrapped with back ticks?

  * **Range of values**: a floating-point number
  * **Type**: `float`
  * **Default value**: `0`
  * **Required**: *No*

**Inputs**:

> *   **1**: `boxes` - tensor of type *T* and shape `[num_batches, num_boxes, 4]` with box coordinates. he box cooridnates are layout as `[xmin, ymin, xmax, ymax]`. Required.

Reviewer: "he box cooridnates are layout as" -> "The box coordinates are layout as"

> *   **1**: `boxes` - tensor of type *T* and shape `[num_batches, num_boxes, 4]` with box coordinates. he box cooridnates are layout as `[xmin, ymin, xmax, ymax]`. Required.

Reviewer: Wrap word "Required" with double "*" characters.

Author: Is there any unified specification about the format? I checked, most Op has no decoration to "Required". Only a small part of Ops decorated to "Required". Single "*" wrap. @lazarevevgeny

*   **2**: `scores` - tensor of type *T* and shape `[num_batches, num_classes, num_boxes]` with box scores. Required.

**Outputs**:

> *   **1**: `selected_outputs` - tensor of type *T_THRESHOLDS* and shape `[number of selected boxes, 6]` containing the selected boxes with score and class as triplets `[class_id, box_score, xmin, ymin, xmax, ymax]`.

Reviewer: "triplets" -> "tuples"

> *   **2**: `selected_indices` - tensor of type *T_IND* and shape `[number of selected boxes, ]` the selected indices in the flattened `boxes`, which are absolute values cross batches.

Reviewer: `[number of selected boxes, ]` -> `[number of selected boxes]`. The same comment for other similar places in the specification.

> *   **2**: `selected_indices` - tensor of type *T_IND* and shape `[number of selected boxes, ]` the selected indices in the flattened `boxes`, which are absolute values cross batches.

Reviewer: "in the flattened `boxes`" -> "in the flattened `boxes` input"

Reviewer: Also add the following statement in the end. "Therefore possible valid values are in the range `[0, num_batches * num_boxes * 4 - 1]`."

*   **3**: `selected_num` - 1D tensor of type *T_IND* and shape `[num_batches, ]` representing the number of selected boxes for each batch element.

> When there is no box selected, `selected_num` will be filled with 0. And both `selected_outputs` and `selected_indices` are null of shape (0,).

Reviewer: "are null of shape (0,)." -> "are empty tensors with one of dimensions equal to `0`."

**Types**

* *T*: floating point type.

* *T_MAX_BOXES*: integer type.

* *T_THRESHOLDS*: floating point type.

* *T_IND*: `int64` or `int32`.

**Example**

```xml
<layer ... type="MatrixNms" ... >

>     <data decay_function='guassian' sort_result="score" output_type="i64"/>

Reviewer: Please, be consistent and use double quotes instead of single quotes.

    <input>
        <port id="0">
            <dim>3</dim>
            <dim>100</dim>
            <dim>4</dim>
        </port>
        <port id="1">
            <dim>3</dim>
            <dim>5</dim>
            <dim>100</dim>
        </port>
        <port id="2"/>
        <port id="3"/>

>         <port id="4"/>

Reviewer: There are only 2 inputs according to the spec.

    </input>
    <output>
        <port id="5" precision="FP32">

>             <dim>-1</dim>

Reviewer: Add a comment here that "-1" means a undefined dimension calculated during the model inference.

##  MatrixNonMaxSuppression<a name="MatrixNonMaxSuppression"></a> {#openvino_docs_ops_sort_MatrixNms_8}

**Versioned name**: *MatrixNonMaxSuppression-8*

**Category**: *Sorting and maximization*

> **Short description**: *MatrixNonMaxSuppression* performs matrix non maximum suppression (NMS) of the boxes with predicted scores.

Reviewer 2: ```suggestion
**Short description**: *MatrixNonMaxSuppression* performs matrix non-maximum suppression (NMS) of the boxes with predicted scores.
```

**Detailed description**: *MatrixNonMaxSuppression* is a multi-phase operation.

> 

Reviewer 2: The operation performs the following:

> 1. First this operation selects candidate bounding boxes with scores higher than `score_threshold`.

Reviewer 2: ```suggestion
1. Selects candidate bounding boxes with scores higher than `score_threshold`.
```

> 2. Second for each class at most `nms_top_k` candidate boxes are selected.

Reviewer 2: ```suggestion
2. For each class, selects at most `nms_top_k` candidate boxes.
```

> 3. Score of the candidate boxes are then decayed according to Matrix NMS algorithm [Wang et al](https://arxiv.org/abs/2003.10152.pdf). This algorithm is applied independently to each class and each batch element. Boxes of `background_class` are skipped and thus eliminated during the process.

Reviewer 2: ```suggestion
3. Decays scores of the candidate boxes according to the Matrix NMS algorithm [Wang et al](https://arxiv.org/abs/2003.10152.pdf). This algorithm is applied independently to each class and each batch element. Boxes of `background_class` are skipped and thus eliminated during the process.
```

> 4. Finanly, the operation selects boxes with the decayed scores higher than `post_threshold`, and selects at most `keep_top_k` scoring candidate boxes per batch element.

Reviewer 2: ```suggestion
4. Selects boxes with the decayed scores higher than `post_threshold`, and selects at most `keep_top_k` scoring candidate boxes per batch element.
```

The Matrix NMS algorithm is described below:
1.  Sort descending the candidate boxes by score, and compute `n*n` pairwise IOU (IntersectionOverUnion) matrix `X` for the top `n` boxes. Suppose `n` is the number of candidate boxes.

> 2.  Set the lower triangle and diagonal of `X` to 0. Therefore get the upper trianglar matrix `X`.

Reviewer 2: ```suggestion
2.  Set the lower triangle and diagonal of `X` to 0. Therefore get the upper triangular matrix `X`.
```

> 3.  Take th column-wise max of `X` to compute a vector `K` of maximun IOU for each candidate box.

Reviewer 2: ```suggestion
3.  Take the column-wise max of `X` to compute a vector `K` of maximum IOU for each candidate box.
```

> 4.  Repeat element value of `K` along axis 1. Suppose this get a matrix `X_cmax`.

Reviewer 2: ```suggestion
4.  Repeat element value of `K` along axis 1. Suppose this gets a matrix `X_cmax`.
```

5.  Compute the decay factor: `decay_factor = exp((X_cmax**2 - X**2) * gaussian_sigma)` if `decay_function` is `guassian`, else `decay_factor = (1 - X) / (1 - X_cmax)`.
6.  Take the column-wise min of `decay_factor`, and element-wise multiply with scores to decay them.

**Attributes**:

* *sort_result*

  * **Description**: *sort_result* specifies the order of output elements.
  * **Range of values**: `class`, `score`, `none`
    * *class* - sort selected boxes by class id (ascending).
    * *score* - sort selected boxes by score (descending).
    * *none* - do not guarantee the order.
  * **Type**: `string`
  * **Default value**: `none`
  * **Required**: *No*

* *sort_result_across_batch*

>   * **Description**: *sort_result_across_batch* is a flag that specifies whenever it is necessary to sort selected boxes across batches or not.

Reviewer 3: PaddlePaddle has an attribute called normalized to specify whether the bounding boxes are normalized by the image width/height. Do we need to add this attribute ?

Author: Don't have to be in OV Op. We will handle it in our PDPD->OV conversion.

Reviewer: How do you know the image size if you want to handle it in PDPD->OV conversion? So I suggest to add this attribute.

Reviewer 3: During op conversion, it gets the attribute value True/False, but during calculation, they have different IOU calculation methods depends on the normalized value. how can we handle this during the Ops conversion ?

Author: This "normalize" attribute is PDPD specific, no need to use image size. It simply add a const offset to the original bbox. And as discussed with ZhangYi, we can handle this with a decomposition. 

Please refer to Paddle2ONNX, which solve this by adding 2 Add operation before NMS. 
https://github.com/PaddlePaddle/Paddle2ONNX/blob/e95695af98f1cf417d3ba2c0e81abf7b59948ab6/paddle2onnx/op_mapper/detection/multiclass_nms.py#L106

Author: @lazarevevgeny I am so sorry I have to reopen this discussion. During test I found it cannot be solved in decomposition. Extra substract operation is required after NMS. It looks awkward. What's more, the extra substracts won't  success as they are dynamic shape input.

>   * **Range of values**: true of false

Reviewer 3: Spell typo should use `or`  instead of `of`

    * *true* - sort selected boxes across batches.
    * *false* - do not sort selected boxes across batches (boxes are sorted per batch element).
  * **Type**: boolean
  * **Default value**: false
  * **Required**: *No*

* *output_type*

  * **Description**: the tensor type of outputs `selected_indices` and `valid_outputs`.
  * **Range of values**: `i64` or `i32`
  * **Type**: `string`
  * **Default value**: `i64`
  * **Required**: *No*

* *score_threshold*

  * **Description**: minimum score to consider box for the processing.
  * **Range of values**: a floating-point number
  * **Type**: `float`
  * **Default value**: `0`
  * **Required**: *No*

* *nms_top_k*

  * **Description**: maximum number of boxes to be selected per class.
  * **Range of values**: an integer
  * **Type**: `int`
  * **Default value**: `-1` meaning to keep all boxes
  * **Required**: *No*

> * *keep_top_k*

Reviewer 3: Notice that the PaddlePaddle has an attribute called normalized, do we have plan to support this ?

  * **Description**: threshold to filter out boxes with low confidence score after decaying.
  * **Range of values**: a floating-point number
  * **Type**: `float`
  * **Default value**: `0`
  * **Required**: *No*

**Inputs**:

*   **1**: `boxes` - tensor of type *T* and shape `[num_batches, num_boxes, 4]` with box coordinates. The box cooridnates are layout as `[xmin, ymin, xmax, ymax]`. **Required.**

*   **2**: `scores` - tensor of type *T* and shape `[num_batches, num_classes, num_boxes]` with box scores. **Required.**

**Outputs**:

*   **1**: `selected_outputs` - tensor of type *T_THRESHOLDS* and shape `[number of selected boxes, 6]` containing the selected boxes with score and class as tuples `[class_id, box_score, xmin, ymin, xmax, ymax]`.

*   **2**: `selected_indices` - tensor of type *T_IND* and shape `[number of selected boxes, 1]` the selected indices in the flattened input `boxes`, which are absolute values cross batches. Therefore possible valid values are in the range [0, num_batches * num_boxes -1].

*   **3**: `selected_num` - 1D tensor of type *T_IND* and shape `[num_batches, ]` representing the number of selected boxes for each batch element.

> When there is no box selected, `selected_num` is filled with 0. `selected_outputs` is an empty tensor of shape (0, 6), and `selected_indices` is an empty tensor of shape (0, 1).

Reviewer: The same comments as for the `MultiClassNMS` spec in https://github.com/openvinotoolkit/openvino/pull/5907

Author: @lazarevevgeny think twice. I am afraid I have to leave shape of selected_indices of [number of selected boxes, 1]. The reason is,

Talked with Baidu, they insist they need shape similar to Paddle.
We cannot solve this by insert Unsqueeze after NMS because of dynamic shape issue. I am afraid Unsqueeze can handle dynamic shape input. For now we have to make sure NMS is the last layer in the ngraph function.

Author: all other change requests are resolved except this shape issue.
